Implement a lookaside storage for signatures of images in Docker registries

mtrmac opened the pull request:
Allows configuring `sigstore` (for RW) and `sigstore-write` (write-only) for images/repositories/namespaces/registries in `atomic.conf`.
    
If configured, uses them to store signatures in the `docker:` transport.

This works fine for `file://` sigstore URLs, for read-only `http://` and `https://` `sigstore` URLs, and is at best a skeleton for writing to `http://` / `https://` (no authentication, to start). So, with this, pushing signatures to remote Docker registries is still not reasonably possible, but at least you can push to a `file://` pointing to a webroot of a sigstore served to the public over HTTP. (Or you can configure `sigstore` to point to HTTP to pull from the same sources as your users, and `sigstore-write` to the webroot.)

Does not deal with any kind of discovery, the lookaside needs to be manually configured in advance.

This pulls in #49 for `types.SystemContext` and significantly extends it, adding a generic `RootForImplicitAbsolutePaths` and adding a `SystemContext` parameter to `types.NewImage{,Source,Destination}` as tentatively discussed in #41 . (Also note that we now implicitly use the `ATOMIC_CONF` environment variable, for compatibility with `projectatomic/atomic`). I will rebase as necessary.

Parsing `atomic.conf` here in generic is admittedly fairly ugly, but perhaps still better than defining an interface for this and requiring `skopeo` to implement that…

More details are in commit messages.

mtrmac commented:
(Tests failing due to API changes, corresponding update is https://github.com/projectatomic/skopeo/issues/170 .)

mtrmac commented:
Rebased, has no dependencies now.

mtrmac commented:
Note that this calls `os.LookupEnv("ATOMIC_CONF")` if the caller does not explicitly set an `atomic.conf` path through `SystemContext.AtomicConfPath` or `SystemContext.RootForImplicitAbsolutePaths`.  Per the conversation in https://github.com/projectatomic/skopeo/issues/178 we should probably allow overriding that.

(Should the default be to use the environment or to ignore it? I am leaning towards “use” but not sure.)

mtrmac commented:
OK, updated with the configuration per https://github.com/projectatomic/atomic/issues/594, now includes documentation and high unit test converage for `lookaside.go`.

@runcom PTAL sometime (_not_ urgent enough to be done over the weekend).

runcom commented:
> If configured, uses them to store signatures in the docker: transport.

should this also be with `docker-daemon:`? just curious and a bit confused maybe

runcom commented on docker/docker_client.go:
```diff
@@ -42,17 +42,17 @@ type dockerClient struct {
-	var registry string
-	if refHostname == dockerHostname {
+// forWriting specifies whether the client will be used for "write" access (in particular passed to lookaside.go:toplevelFromSection)
+func newDockerClient(ctx *types.SystemContext, ref dockerReference, forWriting bool) (*dockerClient, error) {
```
can we name this just `write bool` and document that it'll be used _for writing_? This is not in the Go idioms where variables,arguments and such are just small names. Maybe `mode` or `mask` and have other values to maybe be future proof?

mtrmac replied:
> can we name this just `write bool` and document that it'll be used _for writing_?

I can’t see that having a short name + “for writing” documentation is any shorter or easier to understand than a self-documenting `forWriting`, but, sure, changed.

> Maybe `mode` or `mask` and have other values to maybe be future proof?

I don’t want to get all Design Patterns on this; the underlying configuration file syntax is _not_ future proof that way, and that is the ABI commitment; even if we turned `mode` into an interface we wouldn’t know what to do with it in `registryNamespace.signatureTopLevel`. The parameter is an internal implementation detail which we can generalize at any time if needed.

runcom replied:
> I don’t want to get all Design Patterns on this; the underlying configuration file syntax is not future proof that way, and that is the ABI commitment; even if we turned mode into an interface we wouldn’t know what to do with it in registryNamespace.signatureTopLevel. The parameter is an internal implementation detail which we can generalize at any time if needed.

yeah I pushed too much there :)

runcom commented on docker/lookaside.go:
```diff
@@ -0,0 +1,194 @@
+
+// configuredSignatureStorageBase reads configuration to find an appropriate signature storage URL for ref and forWriting.
+func configuredSignatureStorageBase(ctx *types.SystemContext, ref dockerReference, forWriting bool) (signatureStorageBase, error) {
+	// FIXME? Loading and parsing the config could be cached across calls.
```
yeah, not blocking this PR though IMO

runcom commented on docker/lookaside.go:
```diff
@@ -0,0 +1,194 @@
+		// Look for a full match.
+		identity := ref.PolicyConfigurationIdentity()
+		if ns, ok := config.Docker[identity]; ok {
+			logrus.Debugf(` Using "docker" namespace %s`, identity)
```
is the empty space at the beginning a typo? here and below in the other logrus calls

mtrmac replied:
This is intentional, to make the debug logs reflect the information flow, e.g.

```
DEBU[0000] Using registries.d directory . for sigstore configuration 
DEBU[0000]  Using "docker" namespace docker.io          
DEBU[0000]   Using file:///var/empty                    
```

runcom replied:
mm ok

runcom commented on docker/docker_image_dest.go:
```diff
@@ -168,12 +174,128 @@ func (d *dockerImageDestination) PutManifest(m []byte) error {
+			return false, err
+		}
+		defer res.Body.Close()
+		if res.StatusCode == http.StatusNotFound {
```
question, is there an API to follow when implementing a remove sigstore? sorry if I missed it but I was just wondering this

runcom replied:
forget this one...

mtrmac replied:
Actually, thanks for pointing this out. Without authentication this is basically pointless; we kind of are defining an API here, but we don’t really want to—e.g. if we decided to later implement WebDAV or something, having the attempt an an implementation above could lead to somebody implementing such a server.

Because we are not truly serious about the above code as an API, let’s just drop it for now.

runcom commented on docker/docker_image_src.go:
```diff
@@ -116,12 +142,77 @@ func (s *dockerImageSource) GetBlob(digest string) (io.ReadCloser, int64, error)
 func (s *dockerImageSource) GetSignatures() ([][]byte, error) {
-	return [][]byte{}, nil
+	if s.c.signatureBase == nil { // Skip dealing with the manifest digest if not necessary.
+		return [][]byte{}, nil
```
you can do `return nil, nil` right? the only thing you can do from a slice is to range over it (and `len(nil) == 0)`) or access it w/o checking its length in which case everything breaks - so, I think it's safe to just return `nil`.

mtrmac replied:
I kind of like returning a real slice in here, as opposed to “no value”. You are right that this is safe.

runcom replied:
> I kind of like returning a real slice in here, as opposed to “no value”. You are right that this is safe.

yeah, only overhead is the slice creation but I believe it shouldn't impact anything here (if the go runtime allocates with `{}` as opposed to allocating when something is actually appended there)

runcom commented:
LGTM

This is awesome @mtrmac

runcom commented:
Forget about the lookaside API doc, I just checked the emails :)

mtrmac commented:
> > If configured, uses them to store signatures in the docker: transport.

> should this also be with `docker-daemon:`? just curious and a bit confused maybe

No; when add a lookaside for `docker-daemon`, it will be a hard-coded directory somewhere in `/var`, with no need to configure it. (Though the structure, and parts of the implementation, may be very similar or shared.)

mtrmac commented on docker/lookaside.go:
```diff
@@ -0,0 +1,194 @@
+		return nil, err
+	}
+	for _, configName := range configNames {
+		configPath := filepath.Join(dirPath, configName)
```
Forgot a `.yaml` suffix check here, will add it.

mtrmac commented:
Updated:

 - `s/forWriting/write/`g + comment changes
 - Dropped the pitiful attempts at writing to `http`/`https`; modified the docs to use `file:///` examples
 - Files without `.yaml` suffix are ignored, + a new test fixture
 - Used `require` instead of `assert` in a few tests to avoid panics

@runcom Do you want to take another look?

runcom commented:
> s/forWriting/write/g + comment changes
> Dropped the pitiful attempts at writing to http/https; modified the docs to use file:/// examples
> Files without .yaml suffix are ignored, + a new test fixture
> Used require instead of assert in a few tests to avoid panics

LGTM

mtrmac commented:
:+1:
